File: odk/examples/DevelopersGuide/FirstSteps/FirstLoadComponent/python/FirstLoadComponent.py
# -*- tab-width: 4; indent-tabs-mode: nil; py-indent-offset: 4 -*-
#
# This file is part of the LibreOffice project.
#
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.
#

import officehelper
import sys
import traceback
from com.sun.star.sheet.CellFlags import FORMULA


def main():
    try:
        remote_context = officehelper.bootstrap()
        if remote_context is None:
            print("ERROR: Could not bootstrap default Office.")
            sys.exit(1)
        srv_mgr = remote_context.getServiceManager()
        desktop = srv_mgr.createInstanceWithContext("com.sun.star.frame.Desktop", remote_context)
        spreadsheet_component = desktop.loadComponentFromURL("private:factory/scalc", "_blank", 0, tuple())
        spreadsheets = spreadsheet_component.getSheets()
        spreadsheets.insertNewByName("MySheet", 0)
        elem_type = spreadsheets.getElementType()
        print(elem_type)
        sheet = spreadsheets.getByName("MySheet")
        cell = sheet.getCellByPosition(0, 0)
        cell.setValue(21)
        cell = sheet.getCellByPosition(0, 1)
        cell.setValue(21)
        cell = sheet.getCellByPosition(0, 2)
        cell.setFormula("=sum(A1:A2)")

        cell.setPropertyValue("CellStyle", "Result")

        spreadsheet_controller = spreadsheet_component.getCurrentController()
        spreadsheet_controller.setActiveSheet(sheet)
        cell.setPropertyValue("VertJustify", "com.sun.star.table.CellVertJustify.TOP")
        formula_cells = sheet.queryContentCells(FORMULA)
        formulas = formula_cells.getCells()
        formula_enum = formulas.createEnumeration()

        while formula_enum.hasMoreElements():
            formula_cell = formula_enum.nextElement()
            print("Formula cell in column " + str(formula_cell.getCellAddress().Column)
                  + ", row " + str(formula_cell.getCellAddress().Row)
                  + " contains " + cell.getFormula())

    except Exception as e:
        print(e)
        traceback.print_exc()
        sys.exit(1)


if __name__ == "__main__":
    main()

# vim: set shiftwidth=4 softtabstop=4 expandtab:

File: odk/examples/DevelopersGuide/FirstSteps/FirstLoadComponent/csharp/FirstLoadComponent.cs
/*
 * This file is part of the LibreOffice project.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

using System;

using com.sun.star.beans;
using com.sun.star.container;
using com.sun.star.frame;
using com.sun.star.lang;
using com.sun.star.sheet;
using com.sun.star.table;
using com.sun.star.uno;

try
{
    XComponentContext xContext = NativeBootstrap.bootstrap();
    if (xContext is null)
        Console.Error.WriteLine("Could not bootstrap office");

    XMultiComponentFactory xServiceManager = xContext.getServiceManager();

    IQueryInterface desktop = xServiceManager
        .createInstanceWithContext("com.sun.star.frame.Desktop", xContext);
    XComponentLoader xComponentLoader = desktop.query<XComponentLoader>();

    PropertyValue[] loadProps = Array.Empty<PropertyValue>();
    XComponent xSpreadsheetComponent = xComponentLoader
        .loadComponentFromURL("private:factory/scalc", "_blank", 0, loadProps);

    XSpreadsheetDocument xSpreadsheetDocument = xSpreadsheetComponent
        .query<XSpreadsheetDocument>();

    XSpreadsheets xSpreadsheets = xSpreadsheetDocument.getSheets();
    xSpreadsheets.insertNewByName("MySheet", 0);
    Type elemType = xSpreadsheets.getElementType();
    Console.WriteLine(elemType.FullName);

    Any sheet = xSpreadsheets.getByName("MySheet");
    XSpreadsheet xSpreadsheet = sheet.castOrDefault<XSpreadsheet>();

    XCell xCell = xSpreadsheet.getCellByPosition(0, 0);
    xCell.setValue(21);
    xCell = xSpreadsheet.getCellByPosition(0, 1);
    xCell.setValue(21);
    xCell = xSpreadsheet.getCellByPosition(0, 2);
    xCell.setFormula("=sum(A1:A2)");

    XPropertySet xCellProps = xCell.query<XPropertySet>();
    xCellProps.setPropertyValue("CellStyle", new Any("Result"));

    XModel xSpreadsheetModel = xSpreadsheetComponent.query<XModel>();
    XController xSpreadsheetController = xSpreadsheetModel.getCurrentController();
    XSpreadsheetView xSpreadsheetView = xSpreadsheetController.query<XSpreadsheetView>();
    xSpreadsheetView.setActiveSheet(xSpreadsheet);

    // Example usage of enum values
    xCellProps.setPropertyValue("VertJustify", new Any(CellVertJustify.TOP));

    // Example usage of PropertyValue structs
    loadProps = new PropertyValue[1]
    {
        new PropertyValue()
        {
            Name = "AsTemplate",
            Value = new Any(true),
        }
    };

    // Uncomment to load a Calc file as template
    //xSpreadsheetComponent = xComponentLoader.loadComponentFromURL(
    //    "file:///c:/temp/DataAnalysys.ods", "_blank", 0, loadProps);

    // Example usage of XEnumerationAccess
    XCellRangesQuery xCellQuery = sheet.castOrDefault<XCellRangesQuery>();
    XSheetCellRanges xFormulaCells = xCellQuery.queryContentCells(CellFlags.FORMULA);
    XEnumerationAccess xFormulas = xFormulaCells.getCells();
    XEnumeration xFormulaEnum = xFormulas.createEnumeration();

    while (xFormulaEnum.hasMoreElements())
    {
        Any formulaCell = xFormulaEnum.nextElement();
        xCell = formulaCell.castOrDefault<XCell>();
        XCellAddressable xCellAddress = xCell.query<XCellAddressable>();
        Console.WriteLine($"Formula cell in column {xCellAddress.getCellAddress().Column}, row {xCellAddress.getCellAddress().Row} contains {xCell.getFormula()}");
    }

    return 0;
}
catch (UnoException e)
{
    Console.Error.WriteLine(e.Message);

    return 1;
}